File: benchmarks/10_matmul/bench.c
/* Benchmark 10: Matrix Multiply — naive 400x400 f64 (A * B = C) */
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(void) {
    clock_t t0 = clock();

    int n = 400;
    double *a = (double *)malloc(n * n * sizeof(double));
    double *b = (double *)malloc(n * n * sizeof(double));
    double *c = (double *)calloc(n * n, sizeof(double));

    for (int i = 0; i < n * n; i++) {
        a[i] = (double)i * 0.000001;
        b[i] = (double)(n * n - i) * 0.000001;
    }

    for (int i = 0; i < n; i++) {
        for (int k = 0; k < n; k++) {
            double aik = a[i * n + k];
            for (int j = 0; j < n; j++) {
                c[i * n + j] += aik * b[k * n + j];
            }
        }
    }

    double trace = 0.0;
    for (int i = 0; i < n; i++) trace += c[i * n + i];

    free(a); free(b); free(c);

    clock_t t1 = clock();
    long long ms = (long long)((double)(t1 - t0) / CLOCKS_PER_SEC * 1000.0);
    printf("%.6f\n", trace);
    printf("TIME_MS:%lld\n", ms);
    return 0;
}

File: benchmarks/10_matmul/bench.rs
// Benchmark 10: Matrix Multiply — naive 400x400 f64 (A * B = C)
use std::time::Instant;

fn main() {
    let t0 = Instant::now();

    let n = 400usize;
    let nn = n * n;
    let mut a = vec![0.0_f64; nn];
    let mut b = vec![0.0_f64; nn];
    let mut c = vec![0.0_f64; nn];

    for i in 0..nn {
        a[i] = i as f64 * 0.000_001;
        b[i] = (nn - i) as f64 * 0.000_001;
    }

    for i in 0..n {
        for k in 0..n {
            let aik = a[i * n + k];
            for j in 0..n {
                c[i * n + j] += aik * b[k * n + j];
            }
        }
    }

    let trace: f64 = (0..n).map(|i| c[i * n + i]).sum();

    let ms = t0.elapsed().as_millis();
    println!("{:.6}", trace);
    println!("TIME_MS:{}", ms);
}
